Gamecreation Fixes

AbirRahaman opened the pull request:
Addressed all of the fixes detailed by mat in the response to the last pull request accept for the following:

- Invite players functionality (pushed back to next sprint)
- Adding a back button to the create game screen (presuming parker took care of this)
- getPlayerGames method to populate the home screen feed (rishi is working on this)

To summarize what was done

- Lots of validation for the create game form (blank name handled, blank description handled, no skill level selected handled, error message if no location is selected or if selected datetime is in the past)
- skill level changed to a spinner with preset options
- datetime changed to epoch 
- create button replaced with add button in the top left corner
- loading screen is being used for while the game is created and saved
- after successful game creation and save the screen routes back to the home screen game feed
- user info is now retrieved using the aws amplify auth api (this is for the create game)
- added comments for the functions in CreateGame
- Cleaned up the code in CreateGame (removed unused code and console logs)

All functionality has been tested and is working as expected

mathewacarter29 commented on screens/SignUp.js:
```diff
@@ -50,6 +53,26 @@ function SignIn({ navigation }) {
+      const player = await DataStore.save(
+        new Player({
+          name: name,
+          skill_level: 5,
```
Should skill_level still be a number? or a string - maybe "Beginner" and we will eventually add a way to change it? Or maybe "Any"?

mathewacarter29 commented on screens/CreateGame.js:
```diff
@@ -0,0 +1,257 @@
+//gets the location for the game based on what the user specified 
+async function getLocation(locationName) {
+  const location = await DataStore.query(Location, (l) => l.name.eq(locationName));
+  return location;
```
Should you be returning location[0] here, like in the getPlayer() function?

mathewacarter29 commented on screens/CreateGame.js:
```diff
@@ -0,0 +1,257 @@
+        gameDescription = "";
+    }
+    //if no location is selected display error message
+    if([gameLocation].includes("")){
```
setLoading(false);

mathewacarter29 commented on screens/CreateGame.js:
```diff
@@ -0,0 +1,257 @@
+    //check if specified datetime is in the past 
+    epochNow = Math.floor(Date.now() / 1000);
+    if(epochDate <= epochNow){
+        setShowError(true);
```
setLoading(false);

mathewacarter29 commented on screens/CreateGame.js:
```diff
@@ -0,0 +1,257 @@
+        return;
+    }
+    setShowError(false);
+    navigation.navigate("HomeScreen")
```
This line should be last in the function. We want the API call to finish before we navigate to a new screen. I also don't know how this effects the flow of execution if we have code that we want to execute AFTER we navigate. Also, after the last API call you have to make, make sure to setLoading(false).

mathewacarter29 commented on screens/CreateGame.js:
```diff
@@ -0,0 +1,257 @@
+  }
+
+  catch (error) {
+    console.log('Error saving game', error);
```
setLoading(false)

mathewacarter29 commented on screens/CreateGame.js:
```diff
@@ -0,0 +1,257 @@
+async function create(){
+    setLoading(true);
+    //convert chosen date to epoch timestamp in seconds
+    epochDate = Math.floor(chosenDate.getTime() / 1000);
```
You can do input validation before the setLoading(true) call. This prevents the need to do API calls before we know if the game is valid or not.

mathewacarter29 commented on screens/CreateGame.js:
```diff
@@ -0,0 +1,257 @@
+    const invitedPlayers = await getAllPlayers();
+    invitedPlayers.push(organizer); // add organizer to invited players so that they get added in GamePlayer
+
+    let newGame = await storeGame(gameName, gameDescription, epochDate, gameSkillLevel, organizer, location);
```
Up to you, but it might be better to store all this info in a javascript object instead of passing in a bunch of parameters.

mathewacarter29 commented on screens/CreateGame.js:
```diff
@@ -0,0 +1,257 @@
+    let newGame = await storeGame(gameName, gameDescription, epochDate, gameSkillLevel, organizer, location);
+
+    if (newGame == null) {
+      console.log("Error creating game");
```
setLoading(false)

mathewacarter29 reviewed (changes requested):
CreateGame.js
1. When pressing Create Game with no fields filled out, loading screen never goes away. Make sure that when you check for errors, in addition to setting the error message, you setLoading(false).
2. When pressing Create Game with no fields filled out, I get an exception "TypeError: Cannot read property 'name' of undefined". I suggest doing all the input validation you need at the very beginning of your create() function. That way you don't accidentally make a request to the DB with blank data.
3. Make the error popup render below the create game button. We don't want to see the button shift downwards when the message pops up. This might become a future store - idk if I like the error message popping up in a way that can shift around other elements. Might make it like the profile button where it pops up above everything else at the very bottom of the screen for like 5 seconds or something. Don't worry about that for now tho - the popup should be below the create button.
4. I don't know if you changed the styling of the Game Feed component or the container or something, but when there is data in the Game Feed the view covers part of the profile button text. Maybe try reordering the stuff in HomeScreen's return statement so the Create Game button (+) is right below the profile button?
5. Whenever you are in a function that is making an API call, right before the first one, do setLoading(true). Then, throughout your code, if there is ever a point where an API call fails, or the function returns before the very end, make sure to setLoading(false). Otherwise, the loading screen will never go away and the user can't do anything.

AbirRahaman commented:
Fixed everything accept bullet 4 because I don't think I just changed any of the home screen styling other than adding the plus button for the create game in the top left corner. I think whatever issues there are with that can be fixed once the game feed screen is working with the actual data from the database.
